lib/client.js

```diff
@@ -262,7 +247,40 @@ class AssetComputeClient extends EventEmitter {
     async process(source, renditions, userData) {
         if (!this._registered) {
             await this.register();
+            const waitTime = process.env.REGISTER_WAIT_TIME_MSEC || DEFAULT_REGISTER_WAIT_TIME_MSEC;
+            console.log(`Registered a new journal. Waiting ${waitTime}s before calling /process`);
```

Comment:
Hmm, IIUC, this _always_ waits on the first `process()`.

However, the delay is only there after you called `/register` for the very _first_ time for an integration and it actually created the IO Events journal & event provider, which take some time to propagate. But that will be the vast minority overall. If your app using this client starts up and uses an existing integration that was registered before, there is no need to wait.

So maybe we push this to the client (e.g. the IT tests) to wait between `register()` and `process()`, and have `process()` throw an appropriate error? Not sure if we have a unique error code returned from `/process` right now.

Reply:
I agree we should push this to the client.

IMO, we should throw an error if `(!this._registered)`. Then the client will have to call `.register()` and deal with the decision to wait or not before calling `/process`. 

I think a nice improvement would be also adding some lightweight function `checkJournalRegistration` so the client can check if the journal is registered so they know how long to wait.

Btw, From my experience in testing, when `/process` is called without registering a journal, the IT tests timeout (since we have retry logic sending IO events up until action deadline)

Reply:
On first glance, 1. sounds good, but if we "no longer" implicitly call register(), then that would break existing usages? you need to call /register to get the journal url. I guess there aren't many, so we could change them all... but it would be a breaking change then and need a major version update.

> From my experience in testing, when /process is called without registering a journal, the IT tests timeout

Hmm, actually `/process` should fail with a 400: https://git.corp.adobe.com/nui/api-process/blob/master/process.js#L101-L106
But this only [checks for the presence of the journal](https://git.corp.adobe.com/nui/api-process/blob/master/lib/registration.js#L40), which might be fast (idk), while the event provider (needed for sending events) might be the one that is actually slow to setup in IO Events.

To be honest, I don't think I have the full picture to best judge... what code is a good example of the current usage? And how would roughly the new IT test look like? (Sorry for asking but a lot of stuff today so don't have that much time researching myself)

Reply:
To answer your question, no it wouldn't break existing usages because I made it backward compatible :). I kept the old `createAssetComputeClient` function:
https://github.com/adobe/asset-compute-client/blob/6e9c23a373b7bb1371c510cc3d41725b608b73d8/lib/client.js#L403-L412
That function calls `.register()` and functions exactly how it did before.

Also, the only two places this libraries used is in meahana and the IT tests:
old usage: 
IT tests: https://git.corp.adobe.com/nui/integration-tests/blob/master/lib/framework.js#L272 (would still work the same with changes since it is backward compatible)
Meahana: https://github.com/adobe/asset-compute-devtool/blob/master/server/src/assetComputeDevTool.js#L199

new usage would be like this:
```js
setUpAssetCompute(register=true) {
      const client = new AssetComputeClient;
      if (register) {
           await client.register();
      }
      return client;
}
```
so for the register tests we don't call register first in the set up.

btw, this merge did not release because there is a bug in the release script when the release message has a `/` in it.

Reply:
@alexkli 

> Hmm, actually /process should fail with a 400: https://git.corp.adobe.com/nui/api-process/blob/master/process.js#L101-L106
> But this only checks for the presence of the journal, which might be fast (idk), while the event provider (needed for sending events) might be the one that is actually slow to setup in IO Events.

You are correct. The situation where I got timeout errors was when I had called register but it was not fully set up yet (event provider was not set up). So call to /process worked fine but the action timed out sending events

Reply:
Ok, great, LGTM 👍